axum-extra/src/extract/form.rs

```diff
@@ -0,0 +1,124 @@
+use axum::{
+    async_trait,
+    body::HttpBody,
+    extract::{
+        rejection::{FailedToDeserializeQueryString, FormRejection, InvalidFormContentType},
+        FromRequest, RequestParts,
+    },
+    BoxError,
+};
+use bytes::Bytes;
+use http::Method;
+use serde::de::DeserializeOwned;
+use std::ops::Deref;
+
+/// Extractor that deserializes `application/x-www-form-urlencoded` requests
+/// into some type.
+///
+/// `T` is expected to implement [`serde::Deserialize`].
+///
+/// # Differences from `axum::extract::Form`
+///
+/// This extractor uses [`serde_html_form`] under-the-hood which supports multi-value items. These
+/// are sent by HTML checkboxes and multi selects. Those values can be collected into a `Vec`.
+///
+/// # Example
+///
+/// ```rust,no_run
+/// use axum_extra::extract::Form;
+/// use serde::Deserialize;
+///
+/// #[derive(Deserialize)]
+/// struct Payload {
+///     #[serde(rename = "value")]
+///     values: Vec<String>,
+/// }
+///
+/// async fn accept_form(Form(payload): Form<Payload>) {
+///     // ...
+/// }
+/// ```
+///
+/// [`serde_html_form`]: https://crates.io/crates/serde_html_form
+#[derive(Debug, Clone, Copy, Default)]
+pub struct Form<T>(pub T);
+
+impl<T> Deref for Form<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+#[async_trait]
+impl<T, B> FromRequest<B> for Form<T>
+where
+    T: DeserializeOwned,
+    B: HttpBody + Send,
+    B::Data: Send,
+    B::Error: Into<BoxError>,
+{
+    type Rejection = FormRejection;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        if req.method() == Method::GET {
+            let query = req.uri().query().unwrap_or_default();
+            let value = serde_html_form::from_str(query)
+                .map_err(FailedToDeserializeQueryString::__private_new::<T, _>)?;
+            Ok(Form(value))
+        } else {
+            if !has_content_type::has_content_type(req, &mime::APPLICATION_WWW_FORM_URLENCODED) {
+                return Err(InvalidFormContentType::default().into());
+            }
+
+            let bytes = Bytes::from_request(req).await?;
+            let value = serde_html_form::from_bytes(&bytes)
+                .map_err(FailedToDeserializeQueryString::__private_new::<T, _>)?;
+
+            Ok(Form(value))
+        }
+    }
+}
+
+mod has_content_type {
+    include!(concat!(
+        env!("CARGO_MANIFEST_DIR"),
+        "/../axum/src/extract/has_content_type.rs"
```

Comment:
I think this will either fail publishing, or fail to build when pulled from crates.io 😄

Reply:
ah thats too bad 😜 I've removed it.